File: leetcode/232.implement-queue-using-stacks.py
# @leet start
class MyQueue:

    def __init__(self):
        self.pushed_raw: list[int] = []
        self.ready: list[int] = []

    def _cycle(self):
        if self.ready:
            return

        while self.pushed_raw:
            self.ready.append(self.pushed_raw.pop())

    def push(self, x: int) -> None:
        self.pushed_raw.append(x)

    def pop(self) -> int:
        self._cycle()

        return self.ready.pop()

    def peek(self) -> int:
        self._cycle()

        return self.ready[-1]

    def empty(self) -> bool:
        self._cycle()

        return not self.ready


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
# @leet end

File: leetcode/232.implement-queue-using-stacks.ts
// @leet start
class MyQueue {
  private readonly in: number[] = [];
  private readonly out: number[] = [];

  private cycleIfNeeded(): void {
    if (!this.out.length) {
      while (this.in.length) {
        this.out.push(this.in.pop()!);
      }
    }
  }

  push(x: number): void {
    this.in.push(x);
  }

  pop(): number {
    this.cycleIfNeeded();

    return this.out.pop()!;
  }

  peek(): number {
    this.cycleIfNeeded();

    return this.out[this.out.length - 1];
  }

  empty(): boolean {
    return !this.in.length && !this.out.length;
  }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */
// @leet end
